File: scripts/check-source-size.mjs
#!/usr/bin/env node

import { readdir, readFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";

const rootIndex = process.argv.indexOf("--root");
const ROOT =
  rootIndex === -1
    ? path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..")
    : path.resolve(process.argv[rootIndex + 1]);
const SOURCE_LIMIT = 300;
// Existing debt is explicit and finite. New files must meet the limit; removing
// an entry is the only way to retire debt, so the check never quietly regresses.
const GRANDFATHERED = new Set(["src/app.rs", "src/chrome.rs", "src/model.rs", "src/pages.rs"]);
const SOURCE_EXTENSIONS = new Set([
  ".js",
  ".jsx",
  ".mjs",
  ".cjs",
  ".ts",
  ".tsx",
  ".vue",
  ".rs",
  ".inc",
]);
const IGNORED = new Set([
  ".agent",
  ".agents",
  ".dev",
  ".git",
  ".tmp",
  "build",
  "coverage",
  "dist",
  "dist-electron",
  "node_modules",
  "release",
  "target",
  "vendor",
]);

async function collect(dir, files = []) {
  for (const entry of await readdir(dir, { withFileTypes: true })) {
    if (entry.isDirectory() && !IGNORED.has(entry.name)) {
      await collect(path.join(dir, entry.name), files);
    } else if (entry.isFile() && SOURCE_EXTENSIONS.has(path.extname(entry.name))) {
      files.push(path.join(dir, entry.name));
    }
  }
  return files;
}

function lineCount(text) {
  const lines = text.split(/\r?\n/);
  return lines.at(-1) === "" ? lines.length - 1 : lines.length;
}

const violations = [];
const debt = [];
for (const file of await collect(ROOT)) {
  const lines = lineCount(await readFile(file, "utf8"));
  const label = path.relative(ROOT, file).replaceAll("\\", "/");
  if (lines > SOURCE_LIMIT) {
    if (GRANDFATHERED.has(label)) debt.push(`${label}: ${lines} lines`);
    else violations.push(`${label}: ${lines} lines (max ${SOURCE_LIMIT})`);
  }
}

if (violations.length) {
  console.error(`source size check failed (${violations.length} file(s))`);
  console.error(violations.sort().join("\n"));
  process.exit(1);
}

console.log(
  `source size check passed (${debt.length} grandfathered file(s) over ${SOURCE_LIMIT} lines)`,
);
if (debt.length) console.log(`baseline debt:\n${debt.sort().join("\n")}`);

File: src/bin/check-source-size.rs
//! Source size gate: no new source file may exceed SOURCE_LIMIT lines.
//! Run: cargo run --bin check-source-size
use std::path::{Path, PathBuf};
use std::process::ExitCode;

const SOURCE_LIMIT: usize = 300;
/// Existing debt is explicit and finite. New files must meet the limit; removing
/// an entry is the only way to retire debt, so the check never quietly regresses.
const GRANDFATHERED: &[&str] = &[
    "src/app.rs",
    "src/chrome.rs",
    "src/model.rs",
    "src/pages.rs",
];
const SOURCE_EXTENSIONS: &[&str] = &[
    ".js", ".jsx", ".mjs", ".cjs", ".ts", ".tsx", ".vue", ".rs", ".inc",
];
const IGNORED: &[&str] = &[
    ".agent",
    ".agents",
    ".dev",
    ".git",
    ".tmp",
    "build",
    "coverage",
    "dist",
    "dist-electron",
    "node_modules",
    "release",
    "target",
    "vendor",
];

fn collect(dir: &Path, files: &mut Vec<PathBuf>) {
    for entry in std::fs::read_dir(dir).into_iter().flatten().flatten() {
        let path = entry.path();
        if path.is_dir() {
            let name = entry.file_name();
            if !IGNORED.contains(&name.to_string_lossy().as_ref()) {
                collect(&path, files);
            }
        } else if SOURCE_EXTENSIONS
            .iter()
            .any(|e| entry.file_name().to_string_lossy().ends_with(e))
        {
            files.push(path);
        }
    }
}

fn line_count(text: &str) -> usize {
    let n = text.lines().count();
    // A trailing newline does not count as an extra line; `str::lines` already
    // skips the empty tail, so `n` matches the JS semantics.
    n
}

fn main() -> ExitCode {
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let mut files = Vec::new();
    collect(&root, &mut files);

    let mut violations = Vec::new();
    let mut debt = Vec::new();
    for file in &files {
        let text = match std::fs::read_to_string(file) {
            Ok(t) => t,
            Err(_) => continue,
        };
        let lines = line_count(&text);
        if lines <= SOURCE_LIMIT {
            continue;
        }
        let label = file
            .strip_prefix(&root)
            .unwrap_or(file)
            .to_string_lossy()
            .replace('\\', "/");
        if GRANDFATHERED.contains(&label.as_str()) {
            debt.push(format!("{label}: {lines} lines"));
        } else {
            violations.push(format!("{label}: {lines} lines (max {SOURCE_LIMIT})"));
        }
    }

    if !violations.is_empty() {
        eprintln!("source size check failed ({} file(s))", violations.len());
        violations.sort();
        for v in &violations {
            eprintln!("{v}");
        }
        return ExitCode::FAILURE;
    }

    println!(
        "source size check passed ({} grandfathered file(s) over {SOURCE_LIMIT} lines)",
        debt.len()
    );
    if !debt.is_empty() {
        debt.sort();
        println!("baseline debt:");
        for d in &debt {
            println!("{d}");
        }
    }
    ExitCode::SUCCESS
}
